src/swap/components/SwapProvider.tsx

```diff
@@ -184,25 +201,117 @@ export function SwapProvider({
           return handleError({ swapError: response });
         }
 
-        onSubmit?.(response);
+        await processSwapTransaction({
+          swapTransaction: response,
+          config,
+          setPendingTransaction,
+          setLoading,
+          sendTransactionAsync,
+          onSuccess,
+        });
+
+        // TODO: refresh balances
       } catch (e) {
-        handleError({ swapError: e as SwapError });
+        const userRejected = (e as BaseError).message.includes(
+          'User rejected the request.',
+        );
+        if (userRejected) {
+          setLoading(false);
+          setPendingTransaction(false);
+          handleError({
+            swapError: {
+              code: USER_REJECTED_ERROR_CODE,
+              error: 'User rejected the request.',
+            },
+          });
+        } else {
+          onError?.(e as SwapError);
+          handleError({ swapError: e as SwapError });
+        }
       } finally {
         setLoading(false);
       }
     },
-    [address, handleError, from.amount, from.token, to.token],
+    [
+      address,
+      config,
+      handleError,
+      from.amount,
+      from.token,
+      sendTransactionAsync,
+      to.token,
+    ],
   );
 
   const value = useValue({
     to,
     from,
     error,
     loading,
+    isTransactionPending,
     handleAmountChange,
     handleToggle,
     handleSubmit,
   });
 
   return <SwapContext.Provider value={value}>{children}</SwapContext.Provider>;
 }
+
+export async function processSwapTransaction({
```

Comment:
Niiice love this, let's have this in `core` as his own file. In this will make this file smaller and easier to mantain.
And ofcourse the tests will be on their own file as well.

Reply:
I would push back/disagree with this because imo `core` should be vanillaJS only and `wagmi` is a dependency in this function

Reply:
Thinking out loud but should we have a `components/utils` for decomposing complex business logic into testable functions, for cases like these? cc @abcrane123

Reply:
We should avoid over deep leveling stuff. as it is hard to find them.

Let's do `swap/utils` instead.